### PR TITLE
SYCL RangePolicy: manually specify workgroup size through chunk size

### PR DESCRIPTION
**Pull request by masterleinad**
In some cases, selecting a custom workgroup size gives significantly better performance than relying on the compiler to choose.

### CODE REVIEW

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (pvelesko) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:66]: replace this with 
```
    const typename Policy::index_type id = item.get_global_id() + m_begin;
```

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (pvelesko) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:116]: global_size must be a multiple of local_size or else the kernel launch fails. Replace with 
```
        const auto launch_range =
            std::ceil(float(actual_range) / wgroup_size) * wgroup_size;
```

**Review by pvelesko — CHANGES_REQUESTED** [review]
With the changes I left in the review, the code runs correctly and now outperforms Kokkos::OpenMPTarget for my example

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:116]: How is this different from what I am doing already? I round `actual` to the next multiple of `wgroup_size`.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:66]: Of course.

**Comment by masterleinad** [comment]
`CUDA-10.1-Clang-Tidy` timing out is clearly unrelated.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (nmm0) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:120]: To clarify -- the purpose of this code is to make sure `launch_range` is a multiple of the workgroup size? Might be worth commenting this so it is clear what's going on here.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (nmm0) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:77]: Nitpick: I personally would prefer something like `m_work_size` because range for me implies begin and end. It's also a bit weird that it's not used in the `sycl::item<1>` overload as one would expect from a variable named `range`, but I get why as it is only used since the functor may be executed past the end of the actual memory range.

**Review by nmm0 — APPROVED** [review]
lgtm pending approval by @pvelesko

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (pvelesko) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:120]: Looks good to me @nmm0

Comment (dalg24) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:120]: Add a comment to capture https://github.com/kokkos/kokkos/pull/4875/files?diff=unified&w=1#r827960591

Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:128]: [](https://github.com/kokkos/kokkos/pull/4875/commits/df1a78be3e7bb00be6fe7d14533f54a4952974a0)[df1a78b](https://github.com/kokkos/kokkos/pull/4875/commits/df1a78be3e7bb00be6fe7d14533f54a4952974a0)

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp]: [](https://github.com/kokkos/kokkos/pull/4875/commits/df1a78be3e7bb00be6fe7d14533f54a4952974a0)[df1a78b](https://github.com/kokkos/kokkos/pull/4875/commits/df1a78be3e7bb00be6fe7d14533f54a4952974a0)

**Comment by masterleinad** [comment]
Retest this please.

**Review by Rombur — APPROVED** [review]

**Review by crtrott — CHANGES_REQUESTED** [review]
We need to have a discussion about this. Note that what you use launch bounds for here, is NOT the intended use, and is NOT the effect you get on either HIP or CUDA. I.e. it does not actually set ranges. If people now start using it for that purpose because of SYCL we potentially create issues. I'd rather have you use chunk_size or something like that.

**Comment by masterleinad** [comment]
> We need to have a discussion about this. Note that what you use launch bounds for here, is NOT the intended use, and is NOT the effect you get on either HIP or CUDA. I.e. it does not actually set ranges. If people now start using it for that purpose because of SYCL we potentially create issues. I'd rather have you use chunk_size or something like that.

Sure, it's slightly abusing the concept. My thought was that it fits close enough to what I'm trying to do here since it's an optimization hint (that is actually taking into account). Also, the values for `LaunchBounds` are normally backend-specific and I don't see any other use for `LaunchBounds` in SYCL (upcoming). Anyway, I'm happy to discuss alternatives. Note that this is for `RangePolicy` where we don't have `chunk_size`.

**Comment by dalg24** [comment]
Actually we do (I had to look it up)
https://github.com/kokkos/kokkos/blob/e765d320fe0d5bbb7ff68b4a8b2bbd81f1d80a0e/core/src/Kokkos_ExecPolicy.hpp#L201-L207

**Comment by masterleinad** [comment]
OK, fair enough.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (crtrott) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:128]: what happens if maxTperB is not set (i.e. -1 I think??

**Review by crtrott — CHANGES_REQUESTED** [review]

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp]: Fixed.

**Comment by crtrott** [comment]
@masterleinad we need some clear documentation for this, since this in particular is different from how OpenMP works, where chunksize is the number of consecutive iterations given to a single thread. However, practically always the only performant choice is 1, so that usage of chunk size is kinda useless anyway and reinterpreting it as something like the block size/work group size is probably not a bad idea. We should consider doing this for the other backends I guess.

**Review by crtrott — APPROVED** [review]

**Pull request merged**
